pkg/controller/garbagecollector/garbagecollector.go

```diff
@@ -209,18 +216,21 @@ func (gc *GarbageCollector) Sync(discoveryClient discovery.ServerResourcesInterf
 			utilruntime.HandleError(fmt.Errorf("failed to sync resource monitors: %v", err))
 			return
 		}
-		// TODO: WaitForCacheSync can block forever during normal operation. Could
-		// pass a timeout channel, but we have to consider the implications of
-		// un-pausing the GC with a partially synced graph builder.
-		if !controller.WaitForCacheSync("garbage collector", stopCh, gc.dependencyGraphBuilder.IsSynced) {
+
+		// We will stop waiting for the cache to be synced either if we detect changes to the set of
+		// deletable resources or if told to stop by stopCh.
+		if !controller.WaitForCacheSync("garbage collector", either(resourcesUpdatedCh, stopCh), gc.dependencyGraphBuilder.IsSynced) {
 			utilruntime.HandleError(fmt.Errorf("timed out waiting for dependency graph builder sync during GC sync"))
 			return
```

Comment:
> once we call resyncMonitors with new resources, we should not unlock workers until we have successfully synced

Totally agree. I think we could keep Sync() working instead of idle when workers are locked, by adding a for loop like:
```
func Sync() {
  wait.Until(func() {
    ...
    // something changed. time to sync
    gc.workerLock.Lock()
    defer gc.workerLock.Unlock()
    for {
      syncAttemptDone := make(chan struct{})
      defer close(syncAttemptDone)
      // polling discovery every 30s
      newResources, resourcesUpdatedCh := listWatchDeletableResources(syncAttemptDone)
      gc.resyncMonitors(newResources)
      controller.WaitForCacheSync( either(resourcesUpdatedCh, stopCh) )
    }
  }, 30s)
}
```